File: backend/app/portal/adapters/mockland.py
"""Demonstration adapter — Mockland tourist visa. MOCK ONLY."""
from __future__ import annotations

from urllib.parse import urlparse

from ..contract import PortalAdapter, register_adapter, DEFAULT_PROHIBITED
from ..mock_portal import MOCK_BASE


class MockPortalDriver:
    """Imperative hooks. In a real adapter these use Playwright against a
    Browserbase page; here they call the in-process MockPortal."""
    def __init__(self, portal):
        self.p = portal

    def register(self, **kw): return self.p.register(**kw)
    def submit_captcha(self, **kw): return self.p.submit_captcha(**kw)
    def verify_email(self, **kw): return self.p.verify_email(**kw)
    def login(self, **kw): return self.p.login(**kw)
    def create_application(self, **kw): return self.p.create_application(**kw)
    def upload_document(self, **kw): return self.p.upload_document(**kw)
    def discover_fee(self, **kw): return self.p.discover_fee(**kw)
    def pay(self, **kw): return self.p.pay(**kw)
    def search_appointments(self, **kw): return self.p.search_appointments(**kw)
    def book_appointment(self, **kw): return self.p.book_appointment(**kw)
    def reschedule_appointment(self, **kw): return self.p.reschedule_appointment(**kw)
    def declare_personally(self, **kw): return self.p.declare_personally(**kw)
    def submit(self, **kw): return self.p.submit(**kw)
    def get_application_state(self, **kw): return self.p.get_application_state(**kw)


def build_mockland_adapter(portal) -> PortalAdapter:
    host = urlparse(MOCK_BASE).netloc
    return register_adapter(PortalAdapter(
        adapter_id="mockland-tourist-v1", adapter_version=1,
        destination_country="Mockland", visa_type="tourist",
        portal_operator="Mockland Bureau of Consular Affairs (MOCK)",
        approved_domains=[host],
        registration_url=f"{MOCK_BASE}/register", login_url=f"{MOCK_BASE}/login",
        application_url=f"{MOCK_BASE}/apply", appointment_url=f"{MOCK_BASE}/appointments",
        required_applicant_fields=["full_name", "email", "passport_number", "nationality", "birth_date"],
        required_documents=["passport", "photo"],
        registration_mappings=[{"field": "email", "selector": "#reg-email"},
                               {"field": "password", "selector": "#reg-password", "sensitive": True}],
        application_mappings=[{"field": "passport_number", "selector": "#app-passport"}],
        captcha_detect="#captcha-challenge",
        password_requirements={"minLength": 12},
        payment_policy="applicant", third_party_payment_policy="applicant",
        appointment_search="calendar", appointment_booking="automated",
        reschedule_policy="automated", representative_submission="applicant",
        personal_declaration_required=True, fee_discovery="page",
        confirmation_extraction="confirmation.referenceNo", receipt_extraction="receipt.receiptNo",
        resume_behavior="re-login; reconcile before any retry",
        rate_limits={"searchMinIntervalMs": 30000, "maxChecksPerDay": 24},
        portal_policy_review_date="2026-07-21",
        production_approval_status="mock", production_enabled=False,
        allowed_actions=["navigate", "read", "fill", "click", "upload"],
        prohibited_actions=list(DEFAULT_PROHIBITED),
        driver=MockPortalDriver(portal),
    ))

File: src/main/portal/adapters/mockland.mjs
// Demonstration adapter: "Mockland" tourist visa, driving the in-process
// MockPortal. This is the template every real country adapter follows. It is
// intentionally MOCK-ONLY (productionApprovalStatus: 'mock', productionEnabled:
// false) — a real portal requires its own adapter, legal review, and approval.
//
// The `driver` object holds the imperative hooks the orchestrator calls. In a
// real adapter these use Playwright against `portal` (a browser page); here
// they call the MockPortal API. The orchestrator itself is portal-agnostic.

import { registerAdapter } from '../adapterContract.mjs'
import { MOCK_BASE } from '../mockPortal.mjs'

export function buildMocklandAdapter(portal) {
  return registerAdapter({
    adapterId: 'mockland-tourist-v1',
    adapterVersion: 1,
    destinationCountry: 'Mockland',
    visaType: 'tourist',
    portalOperator: 'Mockland Bureau of Consular Affairs (MOCK)',
    approvedDomains: [new URL(MOCK_BASE).host],
    registrationUrl: `${MOCK_BASE}/register`,
    loginUrl: `${MOCK_BASE}/login`,
    applicationUrl: `${MOCK_BASE}/apply`,
    paymentDomains: [new URL(MOCK_BASE).host],
    appointmentUrl: `${MOCK_BASE}/appointments`,
    supportedLanguages: ['en'],
    requiredApplicantFields: ['fullName', 'email', 'passportNumber', 'nationality', 'birthDate'],
    requiredDocuments: ['passport', 'photo'],
    registrationMappings: [
      { field: 'email', selector: '#reg-email' },
      { field: 'password', selector: '#reg-password', sensitive: true },
      { field: 'fullName', selector: '#reg-name' }
    ],
    loginMappings: [
      { field: 'email', selector: '#login-email' },
      { field: 'password', selector: '#login-password', sensitive: true }
    ],
    applicationMappings: [
      { field: 'passportNumber', selector: '#app-passport' },
      { field: 'nationality', selector: '#app-nationality' },
      { field: 'birthDate', selector: '#app-dob' }
    ],
    allowedActions: ['navigate', 'read', 'fill', 'click', 'upload'],
    prohibitedActions: [
      'solve_captcha', 'bypass_bot_detection', 'spoof_fingerprint', 'rotate_proxy',
      'evade_rate_limit', 'bypass_waiting_room', 'access_other_account', 'intercept_otp',
      'accept_declaration', 'navigate_unapproved_domain'
    ],
    emailVerification: 'link',
    otp: 'none',
    captcha: { detect: '#captcha-challenge' },
    passwordRequirements: { minLength: 12, requireUpper: true, requireDigit: true, requireSymbol: true, maxLength: 64 },
    // The mock portal accepts a controlled card, but third-party payment policy
    // stays 'applicant' so the default path is the applicant-controlled window
    // unless Stripe Issuing is explicitly configured AND approved.
    paymentPolicy: 'applicant',
    thirdPartyPaymentPolicy: 'applicant',
    appointmentSearch: 'calendar',
    appointmentBooking: 'automated', // permitted for this mock when applicant authorizes
    reschedulePolicy: 'automated',
    cancellationRisks: 'Rescheduling frees the current slot only after the new slot is confirmed; a failed reschedule keeps the original.',
    representativeSubmission: 'applicant', // final submit requires the personal declaration
    personalDeclarationRequired: true,
    feeDiscovery: 'page',
    paymentSuccessSignal: 'receipt.receiptNo',
    appointmentSuccessSignal: 'appointment.confirmationNo',
    submissionSuccessSignal: 'confirmation.referenceNo',
    confirmationExtraction: 'confirmation.referenceNo',
    receiptExtraction: 'receipt.receiptNo',
    resumeBehavior: 'Re-establish session via login; reconcile application state before any payment/booking/submit retry.',
    knownFailureStates: ['MAINTENANCE', 'SESSION_EXPIRED', 'CAPTCHA_FAILED', 'PAYMENT_DECLINED', 'REQUIRES_3DS', 'SLOT_TAKEN'],
    rateLimits: { searchMinIntervalMs: 30_000, maxChecksPerDay: 24 },
    portalPolicyReviewDate: '2026-07-21',
    productionApprovalStatus: 'mock',
    productionEnabled: false,

    // Imperative hooks — the only place that knows how to talk to THIS portal.
    driver: {
      async register({ email, password, fullName }) { return portal.register({ email, password, fullName }) },
      async submitCaptcha(args) { return portal.submitCaptcha(args) },
      async verifyEmail(args) { return portal.verifyEmail(args) },
      async login(args) { return portal.login(args) },
      async createApplication(args) { return portal.createApplication(args) },
      async uploadDocument(args) { return portal.uploadDocument(args) },
      async discoverFee(args) { return portal.discoverFee(args) },
      async pay(args) { return portal.pay(args) },
      async complete3ds(args) { return portal.complete3ds(args) },
      async searchAppointments(args) { return portal.searchAppointments(args) },
      async bookAppointment(args) { return portal.bookAppointment(args) },
      async rescheduleAppointment(args) { return portal.rescheduleAppointment(args) },
      async declarePersonally(args) { return portal.declarePersonally(args) },
      async submit(args) { return portal.submit(args) },
      async getApplicationState(args) { return portal.getApplicationState(args) }
    }
  })
}
